### DISCUSSION TITLE
Nested Machines

### DISCUSSION THREAD

**Discussion by Richard57** [adrian-thurston/ragel#84] [Q&A]
I would like to have a regular expression engine that can handle regular languages of Unicode characters that are closed under canonical equivalence.  I would like the languages handled to included the closures of concatenations of languages and, where regular, closures of Kleene star.

I realise that this may be a tall order for Ragel, so I am prepared to generically define the closure of a concatenation from the automata recognising the constituent languages.  However, I am not sure that such a construction can be defined in Ragel.  I have a working implementation of my own, but it is slow and has bugs around the edges.  Most of the states of the NFA are the direct product of the automata of the constituents and a supervisory automation that has 2^55 states for unnormalised input, but 55 for normalised input.  According to its canonical combing class, and incoming letter may be used to update the first coordinate, the second coordinate, or independently both.  The incoming letter also updates the supervisory automaton when it updates the second coordinate.  For the regular expression _/â.*/_, handled as _/â/ . /.*/_, only a handful of states should remain after minimisation and conversion to a DFA.  Assuming it can cope, Ragel may be a good tool for cleaning up and displaying these NFAs.  But, my question is, can Ragel handle the construction of the NFA for the NFD input?

I might still be able to use Ragel indirectly, by reading its machine state tables, constructing the NFA programmatically, and then using Ragel to simplify and display.  That solution is a bit clunky.  At present, my regular engine only constructs the transitions from a state when I have entered it.

**Comment by adrian-thurston** [comment]
As far as I understand your question, you can do that. You can define machines for unicode characters, then use those definitions in the construction of regular languages. You can define machines using UTF-8, 16, or 32. You just need to set the ragel alphabet type appropriately. Then use hex encoding to define the machines. For example, tophat can be:

```
tophat = 0xF0 0x9F 0x8E 0xA9;
```
in UTF-8. Use alphabet type `unsigned char`.

Or if you use UTF-32, with alphabet type `unsigned int` (on a platform where that is 32 bits)

```
tophat = 0x0001F3A9;
```

Then you can write:

```
some_tophats = tophat . tophat*;
```
